instrumentation/http-url-connection/javaagent/src/test/groovy/HttpUrlConnectionTest.groovy

```diff
@@ -258,4 +258,69 @@ class HttpUrlConnectionTest extends HttpClientTest<HttpURLConnection> implements
       }
     }
   }
+
+  def "sun.net.www.protocol.http.HttpURLConnection.getOutputStream should transform GET into POST"() {
+    setup:
+    def url = resolveAddress("/success").toURL()
+    runWithSpan("someTrace") {
+
+      HttpURLConnection connection = url.openConnection()
+
+      def connectionClassName = connection.getClass().getCanonicalName()
+
+      assert "sun.net.www.protocol.http.HttpURLConnection".equals(connectionClassName)
```

Comment:
Will that work on non-OpenJDK JVMs? Can you add some assumption that will skip this test if this class is not present?

Reply:
Nowadays everybody uses class library from openjdk so I wouldn't worry about this class being missing. Though I wouldn't probably build the instrumentation around `sun.net.www.protocol.http.HttpURLConnection` but rather instrument `getOutpuStream` in all subclasses of `java.net.HttpURLConnection` because for this use case, as far as I can tell, it isn't really important on which class `getOutpuStream` is called. Only the fact that it was called matters. Handling this in other classes besides `sun.net.www.protocol.http.HttpURLConnection` could come handy on weblogic where there is a custom url protocol handler (system property `java.protocol.handler.pkgs` can be used to change url protocol handlers).

Reply:
>  Though I wouldn't probably build the instrumentation around sun.net.www.protocol.http.HttpURLConnection but rather instrument getOutpuStream in all subclasses of java.net.HttpURLConnection because for this use case, as far as I can tell, it isn't really important on which class getOutpuStream is called. 

As far as I understand, the #4597 issue was created for the `sun.net.www.protocol.http.HttpURLConnection` class. This issue aims at reflecting the [`getOutputStream()` behavior](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/2544d2a351eca1a3d62276f969dd2d95e4a4d2b6/jdk/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1298) of [the  `sun.net.www.protocol.http.HttpURLConnection` class, extending `java.net.HttpURLConnection`  abstract class](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/2544d2a351eca1a3d62276f969dd2d95e4a4d2b6/jdk/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L92). Could we be sure that all classes extending `java.net.HttpURLConnection` will implement a `getOutputStream()` method that will transform the GET to POST, as in the `sun.net.www.protocol.http.HttpURLConnection` class?

Reply:
> As far as I understand, the #4597 issue was created for the `sun.net.www.protocol.http.HttpURLConnection` class. This issue aims at reflecting the [`getOutputStream()` behavior](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/2544d2a351eca1a3d62276f969dd2d95e4a4d2b6/jdk/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1298) of [the `sun.net.www.protocol.http.HttpURLConnection` class, extending `java.net.HttpURLConnection` abstract class](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/2544d2a351eca1a3d62276f969dd2d95e4a4d2b6/jdk/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L92). Could we be sure that all classes extending `java.net.HttpURLConnection` will implement a `getOutputStream()` method that will transform the GET to POST, as in the `sun.net.www.protocol.http.HttpURLConnection` class?

This is a good question. I believe we can check the value of `getRequestMethod` after calling `getOutputStream`, if it returns `POST` then it was a `POST` request even if it returned `GET` before calling `getOutputStream`.

Reply:
> This is a good question. I believe we can check the value of getRequestMethod after calling getOutputStream, if it returns POST then it was a POST request even if it returned GET before calling getOutputStream.

I am not sure to have been rather explicit. If I was misunderstood., my question was about the expected behavior, not about implementation.

> we can check the value of getRequestMethod after calling getOutputStream, if it returns POST then it was a POST request even if it returned GET before calling getOutputStream.

The [getOutputStream() method of sun... class updates the method field](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/2544d2a351eca1a3d62276f969dd2d95e4a4d2b6/jdk/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1327). So, the getRequestMethod() should return POST even if the HTTP call was done with a GET method.

In term of expected behavior, do we want all classes extending `java.net.HttpURLConnection` to have an instrumentation behavior for the `getOutputStream()` method similar to this of `sun.net.www.protocol.http.HttpURLConnection`? We can't be sure that all classes extending `java.net.HttpURLConnection` implement the `getOutputStream()` method the same way as the `sun.net.www.protocol.http.HttpURLConnection` does (that is to say, changing GET into POST). As in #4597, it may be disturbing not to have instrumentation aligned with the code behavior?

Reply:
2. If you believe it is best to keep the checks for `sun.net.www.protocol.http.HttpURLConnection` then that is ok.
3. Nice catch about the span name. I think we should change it as it too. You could use
```
Span span = Span.fromContext(context);
span.udateName("HTTP " + requestMethod);
```
to update the span name too. You could update the span name at the same place where you change the value for `HTTP_METHOD` attribute.

Reply:
Code updated to set the span name, thank you for the trick